Unable to create wildcard certificates

**Is this a BUG REPORT or FEATURE REQUEST?**:  BUG REPORT

> Uncomment only one, leave it on its own line: 
>
 /kind bug
> /kind feature


**What happened**: I tried creating wildcard certificates using image `canary`. Cert-manager panics while creating an order for the wildcard certificate. However, I got the certificate generated for another ingress which request for the same domain `mydomain.ooo` (no wildcard).

**What you expected to happen**: Order gets completed and new wildcard certificate is created.

**How to reproduce it (as minimally and precisely as possible)**:  I installed cert-manager with static deployment files from : [here](https://github.com/jetstack/cert-manager/blob/master/contrib/manifests/cert-manager/without-rbac.yaml). I am using `quay.io/jetstack/cert-manager-controller:v0.3.1` cert-manager image. Then I created ingress using the following config:
```
{
  "kind": "Ingress",
  "apiVersion": "extensions/v1beta1",
  "metadata": {
    "name": "sub-es",
    "namespace": "development",
    "annotations": {
      "certmanager.k8s.io/acme-challenge-type": "dns01",
      "certmanager.k8s.io/acme-dns01-provider": "cloudflare",
      "certmanager.k8s.io/issuer": "letsencrypt-staging",
      "kubernetes.io/ingress.class": "nginx",
      "kubernetes.io/tls-acme": "true"
    }
  },
  "spec": {
    "tls": [
      {
        "hosts": [
          "*.mydomain.ooo"
        ],
        "secretName": "es-cert"
      }
    ],
    "rules": [
      {
        "host": "es.mydomain.ooo",
        "http": {
          "paths": [
            {
              "path": "/",
              "backend": {
                "serviceName": "myservice",
                "servicePort": 9200
              }
            }
          ]
        }
      }
    ]
  }
}
```


**Anything else we need to know?**: Following are the logs from cert-manager pod:
```
I0705 05:39:05.888329       1 sync.go:126] Certificate "es-cert" for ingress "sub-es" is up to date
I0705 05:39:05.888346       1 controller.go:166] ingress-shim controller: Finished processing work item "development/sub-es"
I0705 05:39:05.894032       1 controller.go:136] issuers controller: syncing item 'cert-manager/letsencrypt-staging'
I0705 05:39:05.894081       1 acme.go:159] getting private key (letsencrypt-staging-&gt;tls.key) for acme issuer cert-manager/letsencrypt-staging
I0705 05:39:05.894446       1 logger.go:67] Calling GetAccount
I0705 05:39:05.894797       1 controller.go:136] issuers controller: syncing item 'development/letsencrypt-staging'
I0705 05:39:05.894825       1 acme.go:159] getting private key (letsencrypt-staging-&gt;tls.key) for acme issuer development/letsencrypt-staging
I0705 05:39:05.895176       1 logger.go:67] Calling GetAccount
I0705 05:39:06.529674       1 setup.go:73] letsencrypt-staging: verified existing registration with ACME server
I0705 05:39:06.529731       1 controller.go:150] issuers controller: Finished processing work item "development/letsencrypt-staging"
I0705 05:39:06.555893       1 setup.go:73] letsencrypt-staging: verified existing registration with ACME server
I0705 05:39:06.555926       1 controller.go:150] issuers controller: Finished processing work item "cert-manager/letsencrypt-staging"
I0705 05:39:07.824772       1 controller.go:177] certificates controller: syncing item 'development/my-wildcard-cert'
I0705 05:39:07.824787       1 controller.go:177] certificates controller: syncing item 'development/es-cert'
I0705 05:39:07.824866       1 sync.go:239] Preparing certificate development/es-cert with issuer
I0705 05:39:07.824875       1 acme.go:159] getting private key (letsencrypt-staging-&gt;tls.key) for acme issuer development/letsencrypt-staging
I0705 05:39:07.825116       1 sync.go:177] Certificate development/my-wildcard-cert scheduled for renewal in 1438 hours
I0705 05:39:07.825136       1 controller.go:191] certificates controller: Finished processing work item "development/my-wildcard-cert"
I0705 05:39:07.825282       1 prepare.go:247] Cleaning up previous order for certificate development/es-cert
I0705 05:39:07.825299       1 prepare.go:263] Cleaning up old/expired challenges for Certificate development/es-cert
I0705 05:39:07.825308       1 logger.go:22] Calling CreateOrder
I0705 05:39:08.267472       1 acme.go:193] Created order for domains: [{dns *.mydomain.ooo}]
I0705 05:39:08.267513       1 logger.go:52] Calling GetAuthorization
I0705 05:39:08.362536       1 logger.go:77] Calling DNS01ChallengeRecord
I0705 05:39:08.362568       1 prepare.go:263] Cleaning up old/expired challenges for Certificate development/es-cert
I0705 05:39:08.362576       1 logger.go:47] Calling GetChallenge
I0705 05:39:08.466710       1 dns.go:78] Checking DNS propagation for "mydomain.ooo" using name servers: [10.0.0.10:53]
I0705 05:39:08.686376       1 dns.go:85] DNS record for "mydomain.ooo" not yet propagated
E0705 05:39:08.686546       1 sync.go:168] [development/es-cert] Error getting certificate 'es-cert': secret "es-cert" not found
E0705 05:39:08.686932       1 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/src/github.com/jetstack/cert-manager/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/go/src/github.com/jetstack/cert-manager/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/github.com/jetstack/cert-manager/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:509
/usr/local/go/src/runtime/panic.go:491
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:367
/go/src/github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1/helpers.go:20
/go/src/github.com/jetstack/cert-manager/pkg/issuer/acme/dns/dns.go:136
/go/src/github.com/jetstack/cert-manager/pkg/issuer/acme/dns/dns.go:67
/go/src/github.com/jetstack/cert-manager/pkg/issuer/acme/prepare.go:236
/go/src/github.com/jetstack/cert-manager/pkg/issuer/acme/prepare.go:173
/go/src/github.com/jetstack/cert-manager/pkg/issuer/acme/prepare.go:146
/go/src/github.com/jetstack/cert-manager/pkg/issuer/acme/prepare.go:94
/go/src/github.com/jetstack/cert-manager/pkg/controller/certificates/sync.go:240
/go/src/github.com/jetstack/cert-manager/pkg/controller/certificates/sync.go:108
/go/src/github.com/jetstack/cert-manager/pkg/controller/certificates/controller.go:215
/go/src/github.com/jetstack/cert-manager/pkg/controller/certificates/controller.go:69
/go/src/github.com/jetstack/cert-manager/pkg/controller/certificates/controller.go:178
/go/src/github.com/jetstack/cert-manager/pkg/controller/certificates/controller.go:183
/go/src/github.com/jetstack/cert-manager/pkg/controller/certificates/controller.go:147
/go/src/github.com/jetstack/cert-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/go/src/github.com/jetstack/cert-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/go/src/github.com/jetstack/cert-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/usr/local/go/src/runtime/asm_amd64.s:2337
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xece4c1]
 goroutine 195 [running]:
github.com/jetstack/cert-manager/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/jetstack/cert-manager/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x111
panic(0x13c8d00, 0x1f699d0)
	/usr/local/go/src/runtime/panic.go:491 +0x283
github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1.(*ACMEIssuerDNS01Config).Provider(0x0, 0xc42057c470, 0xa, 0x1f6a550, 0x0, 0x1f7c7b0)
	/go/src/github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1/helpers.go:20 +0x51
github.com/jetstack/cert-manager/pkg/issuer/acme/dns.(*Solver).solverForIssuerProvider(0xc4205f2930, 0xc42057c470, 0xa, 0x134d380, 0xc42005cef0, 0xc4204ad0e0, 0x1f145a0)
	/go/src/github.com/jetstack/cert-manager/pkg/issuer/acme/dns/dns.go:136 +0x7a
github.com/jetstack/cert-manager/pkg/issuer/acme/dns.(*Solver).Present(0xc4205f2930, 0x1f145a0, 0xc420afab00, 0xc4202e0680, 0xc42058ae70, 0x6a, 0xc4209be600, 0x5b, 0xc42053fbb0, 0x6, ...)
	/go/src/github.com/jetstack/cert-manager/pkg/issuer/acme/dns/dns.go:67 +0x5a
github.com/jetstack/cert-manager/pkg/issuer/acme.(*Acme).presentChallenge(0xc42005f080, 0x1f145a0, 0xc420afab00, 0x1f1c300, 0xc4209aa970, 0xc4202e0680, 0xc42058ae70, 0x6a, 0xc4209be600, 0x5b, ...)
	/go/src/github.com/jetstack/cert-manager/pkg/issuer/acme/prepare.go:236 +0x3a1
github.com/jetstack/cert-manager/pkg/issuer/acme.(*Acme).processChallenge(0xc42005f080, 0x1f145a0, 0xc420afab00, 0x1f1c300, 0xc4209aa970, 0xc4202e0680, 0xc42058ae70, 0x6a, 0xc4209be600, 0x5b, ...)
	/go/src/github.com/jetstack/cert-manager/pkg/issuer/acme/prepare.go:173 +0xb6
github.com/jetstack/cert-manager/pkg/issuer/acme.(*Acme).presentOrder(0xc42005f080, 0x1f145a0, 0xc420afab00, 0x1f1c300, 0xc4209aa970, 0xc4202e0680, 0xc420476d20, 0x159b3c8, 0x30)
	/go/src/github.com/jetstack/cert-manager/pkg/issuer/acme/prepare.go:146 +0x928
github.com/jetstack/cert-manager/pkg/issuer/acme.(*Acme).Prepare(0xc42005f080, 0x1f145a0, 0xc420afab00, 0xc4202e0680, 0x2, 0xc420467998)
	/go/src/github.com/jetstack/cert-manager/pkg/issuer/acme/prepare.go:94 +0x3df
github.com/jetstack/cert-manager/pkg/controller/certificates.(*Controller).issue(0xc4203ee000, 0x1f145a0, 0xc420afab00, 0x1f14860, 0xc42005f080, 0xc4202e0680, 0x0, 0x1ef7920)
	/go/src/github.com/jetstack/cert-manager/pkg/controller/certificates/sync.go:240 +0x163
github.com/jetstack/cert-manager/pkg/controller/certificates.(*Controller).Sync(0xc4203ee000, 0x1f145a0, 0xc420afab00, 0xc4200d2488, 0x1ef7920, 0xc42030e990)
	/go/src/github.com/jetstack/cert-manager/pkg/controller/certificates/sync.go:108 +0x385
github.com/jetstack/cert-manager/pkg/controller/certificates.(*Controller).processNextWorkItem(0xc4203ee000, 0x1f145a0, 0xc420afab00, 0xc4200170a0, 0x13, 0x2, 0x2)
	/go/src/github.com/jetstack/cert-manager/pkg/controller/certificates/controller.go:215 +0x2b5
github.com/jetstack/cert-manager/pkg/controller/certificates.(*Controller).(github.com/jetstack/cert-manager/pkg/controller/certificates.processNextWorkItem)-fm(0x1f145a0, 0xc420afab00, 0xc4200170a0, 0x13, 0x2, 0x134d380)
	/go/src/github.com/jetstack/cert-manager/pkg/controller/certificates/controller.go:69 +0x52
github.com/jetstack/cert-manager/pkg/controller/certificates.(*Controller).worker.func1(0xc4203ee000, 0xc420554da8, 0xc4203021e0, 0x134d380, 0xc42052a1c0, 0x0, 0x0)
	/go/src/github.com/jetstack/cert-manager/pkg/controller/certificates/controller.go:178 +0x24f
github.com/jetstack/cert-manager/pkg/controller/certificates.(*Controller).worker(0xc4203ee000, 0xc4203021e0)
	/go/src/github.com/jetstack/cert-manager/pkg/controller/certificates/controller.go:183 +0x166
github.com/jetstack/cert-manager/pkg/controller/certificates.(*Controller).Run.func1()
	/go/src/github.com/jetstack/cert-manager/pkg/controller/certificates/controller.go:147 +0x33
github.com/jetstack/cert-manager/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc4205e7820)
	/go/src/github.com/jetstack/cert-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x5e
github.com/jetstack/cert-manager/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc4205e7820, 0x3b9aca00, 0x0, 0x1, 0xc4203021e0)
	/go/src/github.com/jetstack/cert-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbd
github.com/jetstack/cert-manager/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc4205e7820, 0x3b9aca00, 0xc4203021e0)
	/go/src/github.com/jetstack/cert-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by github.com/jetstack/cert-manager/pkg/controller/certificates.(*Controller).Run
	/go/src/github.com/jetstack/cert-manager/pkg/controller/certificates/controller.go:147 +0x182
```

**Environment**: I am using AKS
- Kubernetes version (use `kubectl version`): 1.9.6
- Cloud provider or hardware configuration**: Azure
- Install tools: 
- Others: using nginx-ingress: `quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.13.0`

--- a/pkg/apis/certmanager/v1alpha1/helpers.go
+++ b/pkg/apis/certmanager/v1alpha1/helpers.go
@@ -10,6 +10,10 @@ import (
 )
 
 func (i *IssuerStatus) ACMEStatus() *ACMEIssuerStatus {
+	// this is an edge case, but this will prevent panics
+	if i == nil {
+		return &ACMEIssuerStatus{}
+	}
 	if i.ACME == nil {
 		i.ACME = &ACMEIssuerStatus{}
 	}
@@ -17,15 +21,21 @@ func (i *IssuerStatus) ACMEStatus() *ACMEIssuerStatus {
 }
 
 func (a *ACMEIssuerDNS01Config) Provider(name string) (*ACMEIssuerDNS01Provider, error) {
+	if a == nil {
+		return nil, fmt.Errorf("issuer does not contain DNS01 configuration for provider named %q", name)
+	}
 	for _, p := range a.Providers {
 		if p.Name == name {
 			return &(*&p), nil
 		}
 	}
-	return nil, fmt.Errorf("provider '%s' not found", name)
+	return nil, fmt.Errorf("issuer does not contain DNS01 configuration for provider named %q", name)
 }
 
 func (a *ACMECertificateConfig) ConfigForDomain(domain string) *ACMECertificateDomainConfig {
+	if a == nil {
+		return &ACMECertificateDomainConfig{}
+	}
 	for _, cfg := range a.Config {
 		for _, d := range cfg.Domains {
 			if d == domain {
@@ -37,6 +47,10 @@ func (a *ACMECertificateConfig) ConfigForDomain(domain string) *ACMECertificateD
 }
 
 func (c *CertificateStatus) ACMEStatus() *CertificateACMEStatus {
+	// this is an edge case, but this will prevent panics
+	if c == nil {
+		return &CertificateACMEStatus{}
+	}
 	if c.ACME == nil {
 		c.ACME = &CertificateACMEStatus{}
 	}
@@ -44,6 +58,10 @@ func (c *CertificateStatus) ACMEStatus() *CertificateACMEStatus {
 }
 
 func (iss *Issuer) HasCondition(condition IssuerCondition) bool {
+	// this is an edge case, but this will prevent panics
+	if iss == nil {
+		return false
+	}
 	if len(iss.Status.Conditions) == 0 {
 		return false
 	}
@@ -87,6 +105,10 @@ func (iss *Issuer) UpdateStatusCondition(conditionType IssuerConditionType, stat
 }
 
 func (iss *ClusterIssuer) HasCondition(condition IssuerCondition) bool {
+	// this is an edge case, but this will prevent panics
+	if iss == nil {
+		return false
+	}
 	if len(iss.Status.Conditions) == 0 {
 		return false
 	}
@@ -130,6 +152,10 @@ func (iss *ClusterIssuer) UpdateStatusCondition(conditionType IssuerConditionTyp
 }
 
 func (crt *Certificate) HasCondition(condition CertificateCondition) bool {
+	// this is an edge case, but this will prevent panics
+	if crt == nil {
+		return false
+	}
 	if len(crt.Status.Conditions) == 0 {
 		return false
 	}

